learn/how-tos/ssl-offloading.md

```diff
@@ -0,0 +1,99 @@
+---
+title: "SSL Offloading in LoadBalancer"
+id: ""
+---
+---
+
+
+SSL offloading is the process of removing the SSL based encryption from incoming traffic that a web server receives to relieve it from decryption of data. 
+
+LoadBalancers will verify SSL and decrypt the data before sending to backend nodes.
+Backend nodes will always receive http trafic with few `X-Forwarded` headers. 
+
+WaveMaker Apps can be deployed in various Web/Application Servers. 
+We will discuss here what needs to be taken care while Deploying WaveMaker App into Tomcat Server. Similar approach needs to be followed for other Servers.
+
+- LoadBalancers will offload ssl and sends http trafic to tomcat.
+- LoadBalancers will add below Headers to tomcat server.
+   - `X-Forwarded-Proto`
+   - `X-Forwarded-For`
+   - `X-Forwarded-Port`
+- Our WaveMaker app checks the request with request.isSecure() to validate whether it is https or not.
+- Tomcat should able to read this headers and trust that its https.
+- Default tomcat installation don't read these headers. So we have configure Tomcat valve to read all these headers.
+- Update server.xml with below snippet, just above  `<Valve>` tag. 
+- `/usr/local/tomcat/conf` is the defult location of `server.xml`.
+
+```shell
+<Valve className="org.apache.catalina.valves.RemoteIpValve"
+remoteIpHeader="X-Forwarded-For"
+requestAttributesEnabled="true"
+protocolHeader="X-Forwarded-Proto"
+protocolHeaderHttpsValue="https"
+portHeader="X-Forwarded-Port" /> 
+```
+- `protocolHeader` will be considered only for few internal ip ranges. We can change them via internalProxies attribute of the Tomcat valve.
+- Default value for internalProxies="10\\.\\d{1,3}\\.\\d{1,3}\\.\\d{1,3}|192\\.168\\.\\d{1,3}\\.\\d{1,3}|169\\.254\\.\\d{1,3}\\.\\d{1,3}|127\\.\\d{1,3}\\.\\d{1,3}\\.\\d{1,3}|172\\.1[6-9]{1}\\.\\d{1,3}\\.\\d{1,3}|172\\.2[0-9]{1}\\.\\d{1,3}\\.\\d{1,3}|172\\.3[0-1]{1}\\.\\d{1,3}\\.\\d{1,3}|0:0:0:0:0:0:0:1|::1"
+
+- We can add to be allowed ip ranges like below.  
+
+```shell
+<Valve className="org.apache.catalina.valves.RemoteIpValve"
+internalProxies="<Updated-IP-Range>"
+remoteIpHeader="X-Forwarded-For"
+requestAttributesEnabled="true"
+protocolHeader="X-Forwarded-Proto"
+protocolHeaderHttpsValue="https"
+portHeader="X-Forwarded-Port" /> 
+```
+
+### Test Headers in WebApp
+Follow below steps to test which headers are received by Webapp while hitting from browser.
```

Comment:
instead of bundling this with the app, can we ask them to copy into public folder of tomcat?

Reply:
In containerized model, its an update to Dockerfile and adding this file into folder where docker build started.
So its difficult.

From point of routing from LB: AppName based routing might be enabled, so there could be no possibility to access public dir of tomcat